accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,215 @@
+However, NuGet cannot easily determine which tool is a NuGet plugin.
+On the other hand, the `tool path` option allows us to install .NET tools as global tools, but with the binaries located in a specified location. This makes it easier to identify and invoke the appropriate tool for NuGet operations.
+- The binaries are installed in a location that we specify while installing the tool.
+- We can invoke the tool from the installation directory by providing the directory with the command name or by adding the directory to the PATH environment variable.
+- One version of a tool is used for all directories on the machine.
+The `tool path` option aligns well with the design of NuGet plugins architecture, making it the recommended approach for installing and executing NuGet plugins.
+
+### Security considerations
+
+The [.NET SDK docs](https://learn.microsoft.com/dotnet/core/tools/global-tools#check-the-author-and-statistics) clearly state, `.NET tools run in full trust. Don't install a .NET tool unless you trust the author`. This is an important consideration for plugin customers when installing NuGet plugins via .NET Tools in the future.
+
+### Technical explanation
+
+Plugins are discovered through a convention-based directory structure, such as the `%userprofile%/.nuget/plugins` folder on Windows. CI/CD scenarios and power users can use environment variables to override this behavior. Note that only absolute paths are allowed when using these environment variables.
+
+- `NUGET_NETFX_PLUGIN_PATHS`: Defines the plugins used by the .NET Framework-based tooling (NuGet.exe/MSBuild.exe/Visual Studio). This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_NETCORE_PLUGIN_PATHS`: Defines the plugins used by the .NET Core-based tooling (dotnet.exe). This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_PLUGIN_PATHS`: Defines the plugins used for the NuGet process, with priority preserved. If this environment variable is set, it overrides the convention-based discovery. It is ignored if either of the framework-specific variables is specified.
+
+I propose the addition of a `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable. This variable will define the plugins, installed as .NET tools, to be used by both .NET Framework and .NET Core tooling. It will take precedence over `NUGET_PLUGIN_PATHS`. 
+
+The plugins specified in the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable will be used regardless of whether the `NUGET_NETFX_PLUGIN_PATHS` or `NUGET_NETCORE_PLUGIN_PATHS` environment variables are set. The primary reason for this is that plugins installed as .NET tools can be executed in both .NET Framework and .NET Core tooling.
+
+If customers prefer to install NuGet plugins as a global tool instead of a tool-path tool, they can set the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable. This variable should point to the location of the .NET Tool executable that the NuGet Client tooling can invoke when needed.
```

Comment:
It seems odd to me that installing a tool globally then requires setting an environment variable to utilize it. 

I encourage us to think about making sure that `dotnet restore` "just works" whether the credential provider tool is installed as a global tool or as a local tool. The [local tool](https://learn.microsoft.com/dotnet/core/tools/global-tools#install-a-local-tool) option is nice because it allows credential management tools to be configured in code and simplifies the user experience

For example, to get started on a "clean machine":
1. Clone my repo (which has a tool-manifest that includes a credential providers)
1. `dotnet tool restore`  --> installs configured tools, including the credential provider
1. `dotnet restore|build` --> nuget locates and utilizes the installed credential provider tool

If the user/project instead chooses to install the credential provider as a global tool, then it's a separate install step (once per user), and the steps to build are reduced to `clone` followed by `dotnet restore|build`. A key thing here is that there's no additional messing around with setting env vars/paths.

Reply:
Ahh... I see that a discusion of local tools is covered at the end in "Future Possibilities". I would like to see it considered now :)

Reply:
For the nuget cred providers, I think that each can return a response equivalent to "I can't help with this URL", so having to specify which provider works with which source may not be required.

I do like the idea of being able to specify a list of cred provider commands to use for a given project/repo. This would be useful for the case were more than one cred provider truly is needed for a project.

Reply:
Each credential provider need only be started once and can be reused for an arbitrary number of feeds.

Reply:
> Each credential provider need only be started once and can be reused for an arbitrary number of feeds.

Yeah, that's the current behavior with https://github.com/NuGet/NuGet.Client/pull/5563 as long as the feeds configured [have same scheme, host and port data](https://github.com/NuGet/NuGet.Client/blob/1845d6bd450a7453d573035371c9fec43683d1ef/src/NuGet.Core/NuGet.Credentials/CredentialService.cs#L229).